agent/src/datafile.rs

```diff
@@ -714,106 +756,93 @@ impl DataFile {
-                // stat doesn't make sense. Use `zfs get`:
-                //
-                //   # zfs get -pH -o value creation <snapshot_name>
-                //   1644441276
-
-                let mut cmd = Command::new("zfs");
-                cmd.arg("get");
-                cmd.arg("-pH");
-                cmd.arg("-o");
-                cmd.arg("value");
-                cmd.arg("creation");
-                cmd.arg(snapshot_name);
+        info!(self.log, "path is {:?}", &path);
 
-                info!(
-                    self.log,
-                    "command {:?} {:?}",
-                    cmd.get_program(),
-                    cmd.get_args()
-                );
-
-                let cmd = cmd.output()?;
+        let dataset =
+            ZFSDataset::new(path.into_os_string().into_string().unwrap())?;
 
-                if !cmd.status.success() {
-                    bail!("stat didn't work!");
-                }
+        info!(self.log, "dataset is {}", dataset.dataset());
 
-                let cmd_stdout = {
-                    let cmd_stdout = String::from_utf8_lossy(&cmd.stdout);
+        let cmd = Command::new("zfs")
+            .arg("list")
+            .arg("-t")
+            .arg("snapshot")
+            .arg("-H")
+            .arg("-o")
+            .arg("name")
+            .arg("-r")
+            .arg(dataset.dataset())
+            .output()?;
 
-                    // Remove newline
-                    let cmd_stdout = cmd_stdout.trim_end().to_string();
+        if !cmd.status.success() {
+            bail!("zfs list snapshots failed!");
+        }
 
-                    cmd_stdout
-                };
+        let mut results = Vec::new();
 
-                info!(self.log, "stdout is {}", &cmd_stdout);
+        let snapshots_stdout = String::from_utf8_lossy(&cmd.stdout);
+        let snapshots_list: Vec<&str> = snapshots_stdout
+            .trim_end()
+            .split('\n')
+            .filter(|x| !x.is_empty())
+            .collect();
 
-                results.push(Snapshot {
-                    name: dir_name,
-                    created: Utc.timestamp_opt(cmd_stdout.parse()?, 0).unwrap(),
-                });
+        for snapshot in snapshots_list {
+            if snapshot == dataset.dataset() {
+                info!(self.log, "skipping {} matches dataset name", snapshot);
+                continue;
+            }
+            info!(self.log, "snapshot is {}", &snapshot);
+            let parts: Vec<&str> = snapshot.split('@').collect();
+            info!(self.log, "parts is {:?}", &parts);
+            if parts.len() != 2 {
+                panic!("bad snapshot name");
             }
+            let snapshot_name = parts[1];
+
+            let cmd = Command::new("zfs")
+                .arg("get")
+                .arg("-pH")
+                .arg("-o")
+                .arg("value")
+                .arg("creation")
+                .arg(snapshot)
+                .output()?;
+
```

Comment:
Do we need a `if !cmd.status.success() {` check here?  Or if we fail, does cmd_stdout just end up
with nothing in it?

Reply:
Error check added in dfb1358